### PR TITLE
feat: expose underlying client instance

### PR DESCRIPTION
**Pull request by xuzuodong**
### 🔗 Linked issue

#210

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme, or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Adds `instance` key to all drivers.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/drivers/azure-app-configuration.ts [hunk 1]
```diff
@@ -68,6 +68,7 @@ export default defineDriver((opts: AzureAppConfigurationOptions = {}) => {
   return {
     name: DRIVER_NAME,
     options: opts,
+    instance: getClient(),
```
Comment (pi0) [inline src/drivers/azure-app-configuration.ts:71]: It should be a getter `getInstance`

**Comment by xuzuodong** [comment]
I'm not able to add proper `instance` to `mongodb` though since it causes a build error because of [this issue of unbuild](https://github.com/unjs/unbuild/issues/303).

**Comment by pi0** [comment]
Feel free to add for mongo too will try to fix build issue

File: src/drivers/azure-app-configuration.ts (hunk 1 above)
Comment (xuzuodong) [reply src/drivers/azure-app-configuration.ts:71]: Then I guess I will change all of the `instance` key to `getInstance`?

Comment (pi0) [reply src/drivers/azure-app-configuration.ts:71]: Yes please

File: src/drivers/cloudflare-kv-http.ts [hunk 2]
```diff
@@ -206,6 +206,7 @@ export default defineDriver<KVHTTPOptions>((opts) => {
   return {
     name: DRIVER_NAME,
     options: opts,
+    getInstance: undefined,
```
Comment (pi0) [inline src/drivers/cloudflare-kv-http.ts:209]:
```suggestion
```

File: src/drivers/fs-lite.ts [hunk 3]
```diff
@@ -40,6 +40,7 @@ export default defineDriver((opts: FSStorageOptions = {}) => {
   return {
     name: DRIVER_NAME,
     options: opts,
+    getInstance: undefined,
```
Comment (pi0) [inline src/drivers/fs-lite.ts:43]:
```suggestion
```

File: src/drivers/fs.ts [hunk 4]
```diff
@@ -55,6 +55,7 @@ export default defineDriver((opts: FSStorageOptions = {}) => {
   return {
     name: DRIVER_NAME,
     options: opts,
+    getInstance: undefined,
```
Comment (pi0) [inline src/drivers/fs.ts:58]:
```suggestion
```

File: src/drivers/github.ts [hunk 5]
```diff
@@ -84,6 +84,7 @@ export default defineDriver<GithubOptions>((_opts) => {
   return {
     name: DRIVER_NAME,
     options: opts,
+    getInstance: undefined,
```
Comment (pi0) [inline src/drivers/github.ts:87]:
```suggestion
```

File: src/drivers/http.ts [hunk 6]
```diff
@@ -25,6 +25,7 @@ export default defineDriver((opts: HTTPOptions) => {
   return {
     name: DRIVER_NAME,
     options: opts,
+    getInstance: undefined,
```
Comment (pi0) [inline src/drivers/http.ts:28]:
```suggestion
```

File: src/drivers/indexedb.ts [hunk 7]
```diff
@@ -1,13 +1,8 @@
...
-  clear,
-  del,
-  keys,
-  createStore,
-  type UseStore,
-} from "idb-keyval";
+import * as IdbKeyval from "idb-keyval";
+import type { UseStore } from "idb-keyval";
+
+const { get, set, clear, del, keys, createStore } = IdbKeyval;
```
Comment (pi0) [inline src/drivers/indexedb.ts:5]: Can you please revert unrelavant changes?

File: src/drivers/overlay.ts [hunk 8]
```diff
@@ -14,6 +14,7 @@ export default defineDriver((options: OverlayStorageOptions) => {
   return {
     name: DRIVER_NAME,
     options: options,
+    getInstance: undefined,
```
Comment (pi0) [inline src/drivers/overlay.ts:17]:
```suggestion
```

File: src/drivers/utils/index.ts [hunk 9]
```diff
@@ -1,11 +1,15 @@
 import type { Driver } from "../..";
 
-type DriverFactory<T> = (opts: T) => Driver;
+type DriverFactory<
+  Opts,
+  Instance extends undefined | (() => any) = undefined,
```
Comment (pi0) [inline src/drivers/utils/index.ts:5]: Instance type can be returned value of instead of function

File: src/drivers/capacitor-preferences.ts [hunk 10]
```diff
@@ -8,13 +8,17 @@ export interface CapacitorPreferencesOptions {
   base?: string;
 }
 
-export default defineDriver<CapacitorPreferencesOptions>((opts) => {
+export default defineDriver<
+  CapacitorPreferencesOptions,
+  () => typeof Preferences
```
Comment (pi0) [inline src/drivers/capacitor-preferences.ts:13]: I think these generics now need to be updated like this?

```suggestion
  Preferences
```

File: test/drivers/utils.ts [hunk 11]
```diff
@@ -7,7 +7,7 @@ export interface TestContext {
 }
 
 export interface TestOptions {
-  driver: Driver;
+  driver: Driver<any>;
```
Comment (pi0) [inline test/drivers/utils.ts:10]: Generic is optional
```suggestion
  driver: Driver;
```

File: src/types.ts [hunk 12]
```diff
@@ -16,9 +16,10 @@ export interface StorageMeta {
 
 export type TransactionOptions = Record<string, any>;
 
-export interface Driver {
+export interface Driver<T = undefined> {
```
Comment (pi0) [inline src/types.ts:19]: For future compatibiliy, drivers might use first generic for Options type so it would be better something like this:

```suggestion
export interface Driver<OptionsT = any, InstanceT = any> {
```

Comment (xuzuodong) [reply src/types.ts]: Improving `Driver` type introduces a bit more type errors and I tried to fix them.

File: src/drivers/utils/index.ts [hunk 13]
```diff
@@ -1,11 +1,13 @@
...
-type DriverFactory<T> = (opts: T) => Driver;
+type DriverFactory<OptionsT, InstanceT> = (
+  opts: OptionsT
+) => Driver<OptionsT, InstanceT>;
 interface ErrorOptions {}
 
-export function defineDriver<T = any>(
-  factory: DriverFactory<T>
-): DriverFactory<T> {
+export function defineDriver<OptionsT = any, InstanceT = any>(
```
Comment (pi0) [inline src/drivers/utils/index.ts:8]: This way, default instance is not specified unless typed wdyt? 

```suggestion
export function defineDriver<OptionsT = any, InstanceT = never>(
```

File: src/drivers/memory.ts [hunk 14]
```diff
@@ -2,12 +2,13 @@ import { defineDriver } from "./utils";
...
 
-export default defineDriver<void>(() => {
+export default defineDriver<void, Map<string, any>>(() => {
   const data = new Map<string, any>();
 
   return {
     name: DRIVER_NAME,
-    options: {},
+    getInstance: () => data,
+    options: undefined,
```
Comment (pi0) [inline src/drivers/memory.ts:11]: Was this change necessary? 

```suggestion
    options: {},
```

File: src/types.ts [hunk 15]
```diff
@@ -16,9 +16,10 @@ export interface StorageMeta {
 
 export type TransactionOptions = Record<string, any>;
 
-export interface Driver {
+export interface Driver<OptionsT = any, InstanceT = any> {
```
Comment (pi0) [inline src/types.ts:19]:
```suggestion
export interface Driver<OptionsT = any, InstanceT = never> {
```

**Review by pi0 — APPROVED** [review]
Beautiful work, thanks!

File: src/drivers/utils/index.ts (hunk 13 above)
Comment (pi0) [reply src/drivers/utils/index.ts]: had to revert in 256c1923653c5cc76ebc258b667e609503f867e4

**Comment by xuzuodong** [comment]
@pi0 Hi! Would it be possible to release a new version soon to include this? I'm eagerly looking forward to using them in my project. Thanks in advance 🫶

**Comment by pi0** [comment]
Surely. it might be delayed for a little longer since is in parallel to other releases.
